src/track-promise.cc

```diff
@@ -0,0 +1,72 @@
+#include "env.h"
+#include "env-inl.h"
+
+namespace node {
+
+using v8::Function;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::Object;
+using v8::Persistent;
+using v8::Value;
+using v8::WeakCallbackData;
+
+typedef void (*FreeCallback)(Local<Object> object, Local<Function> fn);
+
+class TrackPromise {
+ public:
+  static TrackPromise* New(Isolate* isolate, Local<Object> object);
+  inline Persistent<Object>* persistent();
+ private:
+  static void WeakCallback(const WeakCallbackData<Object, TrackPromise>&);
+  inline void WeakCallback(Isolate* isolate, Local<Object> object);
+  inline TrackPromise(Isolate* isolate, Local<Object> object);
+  ~TrackPromise();
+  Persistent<Object> persistent_;
+};
+
+
+TrackPromise* TrackPromise::New(Isolate* isolate,
+                                Local<Object> object) {
+  return new TrackPromise(isolate, object);
+}
+
+
+Persistent<Object>* TrackPromise::persistent() {
+  return &persistent_;
+}
+
+
+TrackPromise::TrackPromise(Isolate* isolate,
+                           Local<Object> object)
+    : persistent_(isolate, object) {
+  persistent_.SetWeak(this, WeakCallback);
+  persistent_.MarkIndependent();
+}
+
+
+TrackPromise::~TrackPromise() {
+  persistent_.Reset();
+}
+
+
+void TrackPromise::WeakCallback(
+    const WeakCallbackData<Object, TrackPromise>& data) {
+  data.GetParameter()->WeakCallback(data.GetIsolate(), data.GetValue());
+}
+
+
+void TrackPromise::WeakCallback(Isolate* isolate, Local<Object> object) {
+  Environment* env = Environment::GetCurrent(isolate);
+  Local<Function> fn = env->promise_unhandled_rejection();
+
+  if (!fn.IsEmpty()) {
+    HandleScope scope(isolate);
+    Local<Value> promise = object.As<Value>();
+    fn->Call(env->context(), Null(env->isolate()), 1, &promise).ToLocalChecked();
```

Comment:
That's a good point. I think the original intent of `fn` is to conditionally abort the process here. It should be do-able in C++.

Also note that with the new style weakness (which becomes the only non-deprecated option with V8 4.9), the promise object is already be GC'd by the time the weak callback is called and you will not have a handle to it.